src/test/java/com/google/starfish/services/FavoriteNoteServiceTest.java

```diff
@@ -0,0 +1,300 @@
+import com.ninja_squad.dbsetup.destination.DataSourceDestination;
+import com.ninja_squad.dbsetup.operation.Operation;
+
+@RunWith(JUnit4.class)
+public class FavoriteNoteServiceTest {
+  private Constants constants = new Constants();
+  private String NOTES = constants.NOTES;
+  private String LABELS = constants.LABELS;
+  private String FAVORITE_NOTES = constants.FAVORITE_NOTES;
+  private String USERS = constants.USERS;
+  private DataSource pool = Constants.pool;
+  private FavoriteNoteService favoriteNoteService = new FavoriteNoteService();
+
+  private final Date today = favoriteNoteService.getDateBasedOnRecency(Recency.TODAY);
+  private final Date thisWeek = favoriteNoteService.getDateBasedOnRecency(Recency.THIS_WEEK);
+  private final Date thisMonth = favoriteNoteService.getDateBasedOnRecency(Recency.THIS_MONTH);
+  private final Date allTime = favoriteNoteService.getDateBasedOnRecency(Recency.ALL_TIME);
+
+  private final String newSchool = "cornell university";
+  private final String newCourse = "cs1300";
+  private final String firstNewUserId = Integer.toString(Integer.parseInt(constants.REFERENCE_USER_ID) + 1);
+  private final String secondNewUserId = Integer.toString(Integer.parseInt(constants.REFERENCE_USER_ID) + 2);
+
+  private static final boolean runTests = Constants.TEST_DB_NAME.equals("starfish_test");
+
+  /** Clear DB and insert reference data */
+  @Before
+  public void prepare() throws Exception {
+    if(!runTests) throw new Exception("Wrong Test Database Name");
+    Operation operation = getBeforeTestOperation();
+    DbSetup dbSetup = new DbSetup(new DataSourceDestination(pool), operation);
+    dbSetup.launch();
+  }
+
+  /** Test to see if we can get all notes favorited by a single user */
+  @Test
+  public void testGettingFavoriteNotesByUserId() throws SQLException {
+    Note[] favoriteNotes = favoriteNoteService.getFavoriteNotesByUserId(pool, firstNewUserId);
+    assertEquals(favoriteNotes.length, 7);
+  }
+
+  /** Test to see if we can get the number of times a particular note has been favorited */
+  @Test
+  public void testGettingNumFavoritesByNoteId() throws SQLException {
+    long numFavoritesOfNoteOne = favoriteNoteService.getNumFavoritesByNoteId(pool, 1);
+    long numFavoritesOfNoteTwo = favoriteNoteService.getNumFavoritesByNoteId(pool, 2);
+    assertTrue(numFavoritesOfNoteOne == 1);
+    assertTrue(numFavoritesOfNoteTwo == 2);
+  }
+
+  /** Test to see if we can correctly retrieve the most trending notes today */
+  @Test
+  public void testGettingTrendingNotesToday() throws SQLException, Exception {
+    Object[][] trendingNotesToday = favoriteNoteService.getTrendingNotesBySchoolOrCourse(pool, Recency.TODAY, null, null);
+    // We should be getting all 7 notes irrespective of whether they are trending or not
+    if (trendingNotesToday.length != 7) throw new Exception("Retrieved wrong number of notes: " + trendingNotesToday.length);
+    int numTrendingNotes = 2;
```

Comment:
(Here and below) I'm also not sure where this random number is coming from.  I suppose you just know off the top of your head that this number should be two ?  It might be helpful if you created your fake data such that you had an array like `TRENDING_NOTES_TODAY = [note, note]`, then `TRENDING_NOTES_THIS_WEEK = [TRENDING_NOTES_TODAY] + [note, note]` or so so that here you can just say `numTrendingNotes = TRENDING_NOTES_TODAY.length` and it would be much clearer

Reply:
This number basically represents how many notes have a row in the favorite_notes table in the given timespan, so like in the `getBeforeTestOperation` method, our operation would make sure 2 notes are favorited "today". It also is making sure that 2 additional notes are favorited in the timespan of "this week", but since "this week" includes "today", the total number of trending notes "this week" is 2 + 2 = 4. Then same thing for "this month" which gets us 6. And then for "all time" there's one additional note favorited beyond the scope of "this month" so the number of trending notes all time becomes 7.

I'm not sure having arrays for the dummy data would be the best move here in terms of efficiency because that would mean we'd have to construct all those notes separately and then retrieve all those fields when we're setting values to insert in the SQL tables, all when we know from the beginning what data needs to be inserted into the tables.

Reply:
What I think I could do is make all these variables constants of the test class and leave a comment explaining that these numbers mean how many notes in total are favorited in the given timespan, so for anybody in the future looking at these tests, they'd know that the correlation of these constants is to the number of favorites being inserted as dummy data into the favorite_notes table in the `getBeforeTestOperation` method.